Fix script properties reload from external editor

Hilderin opened the pull request:
- Fixes #91035

This fixes the issue where the properties of a custom resource linked to a script are not updated when the script is modified outside the Godot Script Editor.

The problem was caused by the fact that `update_exports` on the script was never called in this situation. I added a call to `update_exports` in `EditorNode::_resources_changed` if the reloaded resource is a script.

While debugging, I also discovered that the global class name cache and the script documentation were not updated when the script was modified in an external editor and the script was never loaded in the editor. This was caused by the `if (ResourceCache::has(path))` condition in `EditorFileSystem::_update_scan_actions`. There's already a similar condition in `EditorNode::_resources_changed` to prevent loading resources that were never loaded.

Additionally, the global class name was never read from the script file when it was modified externally. In `EditorFileSystem::_update_scan_actions`, when processing `ACTION_FILE_RELOAD`, the `FileInfo` was updated, but with the value in memory. To fix this, I called `update_files` with all the reloaded files. This method is already used when scripts are updated in the Script Editor.

https://github.com/user-attachments/assets/180456ad-ff9b-42df-ac59-0b84cddb588e

Hilderin committed:
Fix script properties reload from external editor

kitbdev commented on editor/editor_node.cpp:
```diff
@@ -964,6 +964,12 @@ void EditorNode::_resources_changed(const Vector<String> &p_resources) {
 	if (changed.size()) {
 		for (Ref<Resource> &res : changed) {
 			res->reload_from_file();
+
+			// Update exports to update the properties in the inspector.
+			Ref<Script> scr = Object::cast_to<Script>(res.ptr());
+			if (scr.is_valid()) {
+				scr->update_exports();
+			}
```
Maybe this should be in `Script::reload_from_file()`.

Hilderin replied:
There's already a` reload_from_file` just above and it was not sufficient.

kitbdev replied:
Sorry, I meant that `update_exports()` should be added into the Script's `reload_from_file` implementation instead of having it here, since every time we reload the script we also want to update the exports.
For instance `ScriptEditor::_close_discard_current_tab` also uses `reload_from_file` and it should call `update_exports` too.

Hilderin commented:
While testing I encounter another problem. If I open the script in an external editor like VSCode that is connected to the LSP server and I modify the script in a third editor like notepad or from a git update, VSCode calls the LSP server with a `didChange` message. At this point `GDScriptTextDocument::sync_script_content` was called which called `EditorFileSystem::get_singleton()->update_file(path);` which updates the last modification date of the script but does not reloads the resource. Since the last modification date of the script is updated in `EditorFileSystem`, it is not reloaded from the scan changes in the `EditorFileSystem`.

In this situation, the resource was never reloaded in the editor. 

I removed the call to `EditorFileSystem::update_file` in which caused the desynchronization between the `EditorFileSystem` and the loaded resources.

Hilderin commented:
Just found another bug while testing #95821 when updating a script from external editor not connected to LSP. The documentation of the script (ex: Documentation on a property with ##) is not updated in the editor. Just pushed a fix for that

Hilderin commented:
Rebased and fixed another bug when reloading script from external editor without being connected via LSP where the documentation was not updated if it was opened in the Script Editor.

akien-mga commented on editor/editor_file_system.cpp:
```diff
@@ -43,6 +43,7 @@
 #include "editor/editor_resource_preview.h"
 #include "editor/editor_settings.h"
 #include "editor/project_settings_editor.h"
+#include "plugins/script_editor_plugin.h"
```
We prefer to use "absolute" include paths for consistency.
```suggestion
#include "editor/plugins/script_editor_plugin.h"
#include "editor/project_settings_editor.h"
```

Hilderin replied:
Wow, I copy/paste from the only place in all Godot code base where the absolute path was not used for `script_editor_plugin.h`!!
I changed it in `C:\Projects\godot\editor\connections_dialog.cpp` also.

KoBeWi commented on editor/editor_file_system.cpp:
```diff
@@ -841,7 +833,7 @@ bool EditorFileSystem::_update_scan_actions() {
 	}
 
-	if (reimports.size()) {
+	if (reimports.size() > 0) {
```
```suggestion
	if (!reimports.is_empty()) {
```

Hilderin replied:
Done!

KoBeWi commented on editor/editor_file_system.cpp:
```diff
@@ -1341,8 +1338,7 @@ void EditorFileSystem::_scan_fs_changes(EditorFileSystemDirectory *p_dir, ScanPr
 			}
-		} else if (ResourceCache::has(path)) { //test for potential reload
-
+		} else {
```
This causes reload attempt for non-loaded Resources (I think?). Why this change?

Hilderin replied:
The problem that this fix is if a script is modified in an external editor without LSP (or just a git pull) and that script was never loaded, it will never be processed and that causes the documentation to not update and the file last modification date was never updated the EditorFileSystem.

KoBeWi replied:
Non-loaded Resources shouldn't need reloading, only updating their last checked time 🤔

Hilderin replied:
One thing that is important to known is that the same check is done in `EditorNode::_resources_changed` to prevent resources that are not already loaded to reload. Also, I don't think it's possible to update du documentation without reloading the script. So, my comprehension is that for script, they will all be reloaded in `EditorFileSystem::_update_script_documentation` but for other resources, they should be reloaded only if they were already in the cache due to the condition in `EditorNode::_resources_changed`. The difference is that in the `resources_reload` signal we will also receive resources that were not already loaded.

I think, after thinking about it, I should put back the condition just to add the path in the `reloads` list. That way, only resources that have been already loaded will be present in the `resources_reload` signal and we still get the documentation updated. Sounds good?

KoBeWi replied:
Yep.

Hilderin replied:
Done!

KoBeWi commented on editor/editor_file_system.cpp:
```diff
@@ -852,6 +846,11 @@ bool EditorFileSystem::_update_scan_actions() {
 		ResourceUID::get_singleton()->update_cache();
 	}
 
+	if (reloads.size() > 0) {
```
```suggestion
	if (!reloads.is_empty()) {
```

Hilderin replied:
Done!

KoBeWi reviewed (approved):
I can't reproduce the bug, but this doesn't seem to break anything and the code looks ok.

akien-mga commented:
Thanks!

ryevdokimov commented:
I'm having trouble narrowing down what specifically in my project is causing this error: https://github.com/godotengine/godot/blob/922ae7e7b5fb6c8733a7c03a3c797773f97d9162/modules/gdscript/gdscript.cpp#L747

But it has something to do with this PR since it only started occurring after this was merged and reverting it prevents it. I'm using a lot of tool scripts which may have something to do with it, and it only occurs when the project is first imported. I'll keep digging before I submit an issue because I don't know how to describe what the problem is yet, but maybe it's something more obvious to the people here.

Zireael07 commented:
@ryevdokimov Please open an issue, comments on closed PRs are rarely seen

ryevdokimov commented:
Looks like someone already did: https://github.com/godotengine/godot/issues/97138
